### PR TITLE
Fix for https://github.com/csarofeen/pytorch/issues/1167

### PR DESCRIPTION
**Pull request by csarofeen**
- Needs repro added to python tests.
- Also builds out type casting, should see if we can test more instances of the cast strings with python tests.

Tagging @rdspring1 for build out of the above 2 points.

### CODE REVIEW

**Comment by t-vi** [comment]
Thank you for the quick fix!

Here is the function I hit this with (intersection over union inspired by maskrcnn), please don't hesitate to take it if it can be useful as a test:

```python
@torch.jit.script
def ratio_iou(x1, y1, w1, h1, x2, y2, w2, h2):
    xi = torch.max(x1, x2)                                # Intersection (yi similarly)
    yi = torch.max(y1, y2)                                # Intersection (yi similarly)
    wi = torch.clamp(torch.min(x1+w1, x2+w2) - xi, min=0)
    hi = torch.clamp(torch.min(y1+h1, y2+h2) - yi, min=0)
    area_i = wi * hi                                      # Area Intersection
    area_u = w1 * h1 + w2 * h2 - wi * hi                  # Area Union
    return area_i / torch.clamp(area_u, min=1e-5)
```

**Comment by csarofeen** [comment]
Sure.

Forgot to cross post: https://github.com/csarofeen/pytorch/issues/1167

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -787,6 +787,58 @@ def t(x: torch.Tensor, y: torch.Tensor, z: float):
...
+        magnitude = 100
+        if (dtype_arg1 in self.int_types):
+            x = torch.randint(-magnitude, magnitude, shape, dtype=dtype_arg1, device="cuda")
+        else:
+            x = torch.randn(shape, dtype=dtype_arg1, device="cuda") * magnitude
+        arg2 = int(0)
+        arg3 = int(magnitude * 0.1)
+
+        def clamp0(x: torch.Tensor, f: int):
+            o = 1. * torch.clamp(x, min=f)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: nitpick: let's not do '1.' in case of peephole optimization.

File: test/test_jit_cuda_fuser.py [hunk 2]
```diff
@@ -787,6 +787,58 @@ def t(x: torch.Tensor, y: torch.Tensor, z: float):
...
+            return o
+        clamp0_jit = torch.jit.script(clamp0)
+        self._run_helper(clamp0_jit, clamp0, x, arg2)
+
+        def clamp1(x: torch.Tensor, f: int, ff: int):
+            o = 1. * torch.clamp(x, min=f, max=ff)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: ditto

File: test/test_jit_cuda_fuser.py [hunk 3]
```diff
@@ -787,6 +787,58 @@ def t(x: torch.Tensor, y: torch.Tensor, z: float):
...
+            return o
+        clamp1_jit = torch.jit.script(clamp1)
+        self._run_helper(clamp1_jit, clamp1, x, arg2, arg3)
+
+        def clamp2(x: torch.Tensor, f: float, ff: int):
+            o = 1. * torch.clamp(x, min=f, max=ff)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: ditto

File: test/test_jit_cuda_fuser.py [hunk 4]
```diff
@@ -787,6 +787,58 @@ def t(x: torch.Tensor, y: torch.Tensor, z: float):
...
+            return o
+        clamp2_jit = torch.jit.script(clamp2)
+        self._run_helper(clamp2_jit, clamp2, x, float(arg2), arg3)
+
+        def clamp3(x: torch.Tensor, f: int, ff: float):
+            o = 1. * torch.clamp(x, min=f, max=ff)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: ditto

File: test/test_jit_cuda_fuser.py [hunk 5]
```diff
@@ -787,6 +787,58 @@ def t(x: torch.Tensor, y: torch.Tensor, z: float):
...
+            return o
+        clamp3_jit = torch.jit.script(clamp3)
+        self._run_helper(clamp3_jit, clamp3, x, arg2, float(arg3))
+
+        def threshold(x: torch.Tensor, th: int, val: int):
+            o = 1. * torch.threshold(x, th, val)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: ditto

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 6]
```diff
@@ -1012,32 +1011,31 @@ Val* clamp(Val* in, Val* min_val, Val* max_val) {
...
-        isFloatingPointType(min_type) && isFloatingPointType(max_type),
-        "All input DataType values should match the input type ",
-        in_type,
-        " vs ",
-        min_type,
-        " and ",
-        max_type);
-  } else if (isIntegralType(in_type)) {
+
     TORCH_CHECK(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: nitpick: indent

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: how's this not picked up by clang-format :confused:

**Review by jjsjann123 — APPROVED** [review]
LGTM, minor comments

**Pull request merged**